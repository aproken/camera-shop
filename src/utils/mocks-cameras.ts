import { Cameras } from '../types/camera';
import { Type, Category, Level, } from '../const';

export const makeFakeCameras = (): Cameras => [
  {
    id: 1,
    name: 'Ретрокамера Dus Auge lV',
    vendorCode: 'DA4IU67AD5',
    type: Type.Collectible,
    category: Category.VideoCamera,
    description: 'Немецкий концерн BRW разработал видеокамеру Das Auge IV в начале 80-х годов, однако она до сих пор пользуется популярностью среди коллекционеров и яростных почитателей старинной техники. Вы тоже можете прикоснуться к волшебству аналоговой съёмки, заказав этот чудо-аппарат. Кто знает, может с Das Auge IV начнётся ваш путь к наградам всех престижных кинофестивалей.',
    previewImg: 'img/content/das-auge.jpg',
    level: Level.Amateur,
    price: 73450,
    previewImg2x: 'img/content/das-auge@2x.jpg',
    previewImgWebp: 'img/content/das-auge.webp',
    previewImgWebp2x: 'img/content/das-auge@2x.webp',
    reviewCount: 117,
    averageRating: 3,
  },
  {
    id: 2,
    name: 'FastShot MR-5',
    vendorCode: 'JH34KHN895',
    type: Type.Momentary,
    category: Category.PhotoCamera,
    description: 'Новое слово в создании моментальных фото. Высокое качество снимков, легко перезаряжаемые кассеты, встроенная вспышка. Создавайте альбомы здесь и сейчас.',
    previewImg: 'img/content/fast-shot.jpg',
    level: Level.Amateur,
    price: 18970,
    previewImg2x: 'img/content/fast-shot@2x.jpg',
    previewImgWebp: 'img/content/fast-shot.webp',
    previewImgWebp2x: 'img/content/fast-shot@2x.webp',
    reviewCount: 77,
    averageRating: 5,
  },
  {
    id: 3,
    name: 'Instaprinter P2',
    vendorCode: 'KLU789GH56',
    type: Type.Digital,
    category: Category.PhotoCamera,
    description: 'Компактная модель позволяющая получать чёткие снимки с 25-кратным зумом. В комплекте зарядное устройство и бархатный чехол, а так же удобный шнурок на шею.',
    previewImg: 'img/content/instaprinter.jpg',
    level: Level.Zero,
    price: 8430,
    previewImg2x: 'img/content/instaprinter@2x.jpg',
    previewImgWebp: 'img/content/instaprinter.webp',
    previewImgWebp2x: 'img/content/instaprinter@2x.webp',
    reviewCount: 128,
    averageRating: 4,
  },
  {
    id: 4,
    name: 'Орлёнок',
    vendorCode: 'O78DFGSD83',
    type: Type.Membranous,
    category: Category.PhotoCamera,
    description: 'Плёночная перезаряжаемая камера нового покаления уже укомплектована плёнкой и оснащена встроенной вспышкой. Легко помещается в руке и обладет интересным дизайном.',
    previewImg: 'img/content/orlenok.jpg',
    level: Level.Amateur,
    price: 19970,
    previewImg2x: 'img/content/orlenok@2x.jpg',
    previewImgWebp: 'img/content/orlenok.webp',
    previewImgWebp2x: 'img/content/orlenok@2x.webp',
    reviewCount: 34,
    averageRating: 2,
  },
  {
    id: 5,
    name: 'Van Shot',
    vendorCode: 'YU7RT5GH76',
    type: Type.Collectible,
    category: Category.VideoCamera,
    description: 'Крайне редкое наименование не потеряло актуальность не смотря на сможество альтернатив. После съёмок на данную камеру фильм не стыдно показать в рамках кинофестиваля. Первые 4К настройки, высочайшее разрешение, уникальная цветопередача.',
    previewImg: 'img/content/van-shot.jpg',
    level: Level.Professional,
    price: 149990,
    previewImg2x: 'img/content/van-shot@2x.jpg',
    previewImgWebp: 'img/content/van-shot.webp',
    previewImgWebp2x: 'img/content/van-shot@2x.webp',
    reviewCount: 32,
    averageRating: 2,
  },
  {
    id: 6,
    name: 'Click Sap',
    vendorCode: 'KLN54H76F5',
    type: Type.Membranous,
    category: Category.PhotoCamera,
    description: 'Зеркальная камера позволяющая делать четкие фотографии. Вспышка продается и подключается отдельно. Чехол в комплекте. Плёнка 35мм',
    previewImg: 'img/content/click-sap.jpg',
    level: Level.Amateur,
    price: 9490,
    previewImg2x: 'img/content/click-sap@2x.jpg',
    previewImgWebp: 'img/content/click-sap.webp',
    previewImgWebp2x: 'img/content/click-sap@2x.webp',
    reviewCount: 23,
    averageRating: 3,
  },
  {
    id: 7,
    name: 'Look 54',
    vendorCode: 'NB54Y',
    type: Type.Digital,
    category: Category.PhotoCamera,
    description: 'Профессиональный зеркальный фотоаппарат оснащен 56-кратным зумом, позволяет создавать чёткие снимки, а новейший процессор позволяет справляться с шумами и светочувствительностью.',
    previewImg: 'img/content/look-54.jpg',
    level: Level.Professional,
    price: 96490,
    previewImg2x: 'img/content/look-54@2x.jpg',
    previewImgWebp: 'img/content/look-54.webp',
    previewImgWebp2x: 'img/content/look-54@2x.webp',
    reviewCount: 31,
    averageRating: 3,
  },
  {
    id: 8,
    name: 'Look SF3',
    vendorCode: 'NBSF3',
    type: Type.Digital,
    category: Category.PhotoCamera,
    description: 'Идеальная камера для старта в репортажной фотографии. оснащена встроенным стабилизатором и вспышкой. Оптический 15 кратный зум и удобная посадка в руке. ',
    previewImg: 'img/content/look-sf3.jpg',
    level: Level.Professional,
    price: 63800,
    previewImg2x: 'img/content/look-sf3@2x.jpg',
    previewImgWebp: 'img/content/look-sf3.webp',
    previewImgWebp2x: 'img/content/look-sf3@2x.webp',
    reviewCount: 39,
    averageRating: 3,
  },
];
