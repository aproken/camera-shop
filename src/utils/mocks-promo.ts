import { Promo } from '../types/promo';

export const makeFakePromo = (): Promo => (
  {
    id: 7,
    name: 'Look 54',
    previewImg: 'img/content/promo.jpg',
    previewImg2x: 'img/content/promo@2x.jpg',
    previewImgWebp: 'img/content/promo.webp',
    previewImgWebp2x: 'img/content/promo@2x.webp'
  }
);
