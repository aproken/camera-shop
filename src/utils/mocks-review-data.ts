import { ReviewData } from '../types/review-data';

export const makeFakeReviewData = (): ReviewData => ({
  cameraId: 1,
  userName: 'Кирилл',
  advantage: 'Легкая в плане веса, удобная в интерфейсе',
  disadvantage: 'Быстро садиться зарядка',
  review: 'Это моя первая камера. Я в восторге, нареканий нет',
  rating: 5,
});
