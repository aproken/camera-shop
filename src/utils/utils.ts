import { SortByType, SortByOrder } from '../const';
import { Cameras } from '../types/camera';
import { Orders } from '../types/order';

//Функция для вычисления количества страниц для пагинации
export const getPageNumbers = (countTotal: number, countOnPage: number) =>
  Array
    .from(
      { length: Math.ceil(countTotal / countOnPage) }, (_value, index) => index + 1);

//Функция для получения товаров на конкретную страницу пагинации
export const getProductsCurrentPage = <T>(products: Array<T>, pageNumberCurrentPage: number, productCountOnPage: number): Array<T> => {
  const endIndex = pageNumberCurrentPage * productCountOnPage;
  const beginIndex = endIndex - productCountOnPage;
  return products.slice(beginIndex, endIndex);
};

//Функция для стилизации цены товара
export const getStylizedPrice = (price: number): string => new Intl.NumberFormat('ru-Ru').format(price);

//Функция для сортировки товаров
export const sortProducts = (
  products: Cameras,
  sortByType: string,
  sortByOrder: string,
): Cameras => {
  // сортировка по цене
  if (sortByType === SortByType.Price) {
    if (sortByOrder === SortByOrder.Down) {
      return [...products].sort((a, b) => b.price - a.price);
    } else {
      return [...products].sort((a, b) => a.price - b.price);
    }
  }

  // Сортировка по популярности (среднему рейтингу)
  if (sortByType === SortByType.Popularity) {
    return [...products].sort((a, b) => {
      if (a.averageRating === null && b.averageRating === null) {
        return 0;
      }
      if (a.averageRating === null) {
        return 1;
      }
      if (b.averageRating === null) {
        return -1;
      }

      const popularityComparison = b.averageRating - a.averageRating;

      if (popularityComparison === 0) {
        if (sortByOrder === SortByOrder.Down) {
          return b.price - a.price;
        } else {
          return a.price - b.price;
        }
      }

      return popularityComparison;
    });
  }

  return products;
};

//Возвращает минимальную цену из списка продуктов
export const getMinPrice = (products: Cameras) => {
  if(products.length !== 0) {
    const prices = products.map((product) => product.price);
    return Math.min(...prices);
  } else {
    return 0;
  }
};

//Возвращает максимальную цену из списка продуктов
export const getMaxPrice = (products: Cameras) => {
  if(products.length !== 0) {
    const prices = products.map((product) => product.price);
    return Math.max(...prices);
  } else {
    return 0;
  }
};

//Возвращает общее количество товаров в корзине
export const getTotalQuantity = (orders: Orders) => {
  if(orders.length) {
    return orders.reduce((totalQuantity, order) => totalQuantity + order.quantity, 0);
  }
};

//Вычисляет итоговую стоимость, учитывая количество товара
export const getTotalPrice = (orders: Orders): number => {
  const total = orders.reduce(
    (totalPrice, order) => totalPrice + order.camera.price * order.quantity,
    0
  );
  return total;
};

//Вычисляет скидку
export const getTotalDiscount = (orders: Orders, discount: number): number => {
  const total = getTotalPrice(orders);
  return Math.floor(total / 100 * discount);
};

//Вычисляет итоговую стоимость, учитывая количество товара и скидку
export const getFinalPrice = (orders: Orders, discount: number): number => {
  const price = getTotalPrice(orders);
  const discountValue = getTotalDiscount(orders, discount);
  return price - discountValue;
};

