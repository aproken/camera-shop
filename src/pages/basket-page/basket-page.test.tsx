import { render } from '@testing-library/react';
import BasketPage from './basket-page';
import { makeFakeCameras } from '../../utils/mocks-cameras';
import thunk from 'redux-thunk';
import { configureMockStore } from '@jedmao/redux-mock-store';
import { Provider } from 'react-redux';
import { createMemoryHistory } from 'history';
import HistoryRouter from '../../components/history-route/history-route';


const mockStore = configureMockStore([thunk]);
const camera = makeFakeCameras()[0];
const order = {
  camera: camera,
  quantity: 1
};
const store = mockStore({
  'ORDER' : {
    ordersList: [order],
    coupon: {
      coupon: 'COUPON_TEXT',
      status: 'error',
      discount: 0
    }
  }
});

describe('BasketPage', () => {
  it('render', () => {
    const history = createMemoryHistory();
    const { container } = render(
      <Provider store={store}>
        <HistoryRouter history={ history }>
          <BasketPage/>
        </HistoryRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
