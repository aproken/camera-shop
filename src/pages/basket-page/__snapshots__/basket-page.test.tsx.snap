// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasketPage render 1`] = `
<div>
  <main
    id="basket-page"
  >
    <div
      class="page-content"
    >
      <div
        class="breadcrumbs"
      >
        <div
          class="container"
        >
          <ul
            class="breadcrumbs__list"
          >
            <li
              class="breadcrumbs__item"
            >
              <a
                class="breadcrumbs__link"
                href="/"
              >
                Главная
                <svg
                  aria-hidden="true"
                  height="8"
                  width="5"
                >
                  <use
                    xlink:href="#icon-arrow-mini"
                  />
                </svg>
              </a>
            </li>
            <li
              class="breadcrumbs__item"
            >
              <a
                class="breadcrumbs__link"
                href="/"
              >
                Каталог
                <svg
                  aria-hidden="true"
                  height="8"
                  width="5"
                >
                  <use
                    xlink:href="#icon-arrow-mini"
                  />
                </svg>
              </a>
            </li>
            <li
              class="breadcrumbs__item"
            >
              <span
                class="breadcrumbs__link breadcrumbs__link--active"
              >
                Корзина
              </span>
            </li>
          </ul>
        </div>
      </div>
      <section
        class="basket"
      >
        <div
          class="container"
        >
          <h1
            class="title title--h2"
          >
            Корзина
          </h1>
          <ul
            class="basket__list"
          >
            <li
              class="basket-item"
            >
              <div
                class="basket-item__img"
              >
                <picture>
                  <source
                    srcset="/img/content/das-auge.webp, /img/content/das-auge@2x.webp"
                    type="image/webp"
                  />
                  <img
                    alt="Ретрокамера Dus Auge lV"
                    height="120"
                    src="/img/content/das-auge.jpg"
                    srcset="/img/content/das-auge@2x.jpg"
                    width="140"
                  />
                </picture>
              </div>
              <div
                class="basket-item__description"
              >
                <p
                  class="basket-item__title"
                >
                  Ретрокамера Dus Auge lV
                </p>
                <ul
                  class="basket-item__list"
                >
                  <li
                    class="basket-item__list-item"
                  >
                    <span
                      class="basket-item__article"
                    >
                      Артикул:
                    </span>
                     
                    <span
                      class="basket-item__number"
                    >
                      DA4IU67AD5
                    </span>
                  </li>
                  <li
                    class="basket-item__list-item"
                  >
                    Коллекционная видеокамера 
                  </li>
                  <li
                    class="basket-item__list-item"
                  >
                    Любительский уровень
                  </li>
                </ul>
              </div>
              <p
                class="basket-item__price"
              >
                <span
                  class="visually-hidden"
                >
                  Цена:
                </span>
                73 450 ₽
              </p>
              <div
                class="quantity"
              >
                <button
                  aria-label="уменьшить количество товара"
                  class="btn-icon btn-icon--prev"
                  disabled=""
                >
                  <svg
                    aria-hidden="true"
                    height="12"
                    width="7"
                  >
                    <use
                      xlink:href="#icon-arrow"
                    />
                  </svg>
                </button>
                <label
                  class="visually-hidden"
                  for="counter1"
                />
                <input
                  aria-label="количество товара"
                  id="counter1"
                  max="99"
                  min="1"
                  type="number"
                  value="1"
                />
                <button
                  aria-label="увеличить количество товара"
                  class="btn-icon btn-icon--next"
                >
                  <svg
                    aria-hidden="true"
                    height="12"
                    width="7"
                  >
                    <use
                      xlink:href="#icon-arrow"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="basket-item__total-price"
              >
                <span
                  class="visually-hidden"
                >
                  Общая цена:
                </span>
                73 450 ₽
              </div>
              <button
                aria-label="Удалить товар"
                class="cross-btn"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  height="10"
                  width="10"
                >
                  <use
                    xlink:href="#icon-close"
                  />
                </svg>
              </button>
            </li>
          </ul>
          <div
            class="basket__summary"
          >
            <div
              class="basket__promo"
            >
              <p
                class="title title--h4"
              >
                Если у вас есть промокод на скидку, примените его в этом поле
              </p>
              <div
                class="basket-form"
              >
                <form
                  action="#"
                >
                  <div
                    class="custom-input is-invalid"
                  >
                    <label>
                      <span
                        class="custom-input__label"
                      >
                        Промокод
                      </span>
                      <input
                        name="promo"
                        placeholder="Введите промокод"
                        type="text"
                        value="COUPON_TEXT"
                      />
                    </label>
                    <p
                      class="custom-input__error"
                    >
                      Промокод неверный
                    </p>
                  </div>
                  <button
                    class="btn"
                    type="submit"
                  >
                    Применить
                  </button>
                </form>
              </div>
            </div>
            <div
              class="basket__summary-order"
            >
              <p
                class="basket__summary-item"
              >
                <span
                  class="basket__summary-text"
                >
                  Всего:
                </span>
                <span
                  class="basket__summary-value"
                >
                  73 450 ₽
                </span>
              </p>
              <p
                class="basket__summary-item"
              >
                <span
                  class="basket__summary-text"
                >
                  Скидка:
                </span>
                <span
                  class="basket__summary-value basket__summary-value--bonus"
                >
                  0  ₽
                </span>
              </p>
              <p
                class="basket__summary-item"
              >
                <span
                  class="basket__summary-text basket__summary-text--total"
                >
                  К оплате:
                </span>
                <span
                  class="basket__summary-value basket__summary-value--total"
                >
                  73 450 ₽
                </span>
              </p>
              <button
                class="btn btn--purple"
                type="submit"
              >
                Оформить заказ
              </button>
            </div>
          </div>
        </div>
      </section>
    </div>
  </main>
</div>
`;
